hpx/parallel/algorithms/transfer.hpp

```diff
@@ -0,0 +1,155 @@
+    ///                     input iterator.
+    /// \tparam OutIter     The type of the iterator representing the
+    ///                     destination range (deduced).
+    ///                     This iterator type must meet the requirements of an
+    ///                     output iterator.
+    ///
+    /// \param policy       The execution policy to use for the scheduling of
+    ///                     the iterations.
+    /// \param first        Refers to the beginning of the sequence of elements
+    ///                     the algorithm will be applied to.
+    /// \param last         Refers to the end of the sequence of elements the
+    ///                     algorithm will be applied to.
+    /// \param dest         Refers to the beginning of the destination range.
+    ///
+    /// \returns  The \a transfer algorithm returns a \a hpx::future<OutIter> if
+    ///           the execution policy is of type
+    ///           \a sequential_task_execution_policy or
+    ///           \a parallel_task_execution_policy and
+    ///           returns \a OutIter otherwise.
+    ///           The \a move algorithm returns the output iterator to the
+    ///           element in the destination range, one past the last element
+    ///           transfered.
+    ///
+    template <
+        template <typename IterPair> class Algo,
+        typename ExPolicy, typename InIter, typename OutIter,
+        HPX_CONCEPT_REQUIRES_(
+            is_execution_policy<ExPolicy>::value &&
+            traits::is_iterator<InIter>::value &&
+            traits::is_iterator<OutIter>::value)
+    >
+    typename util::detail::algorithm_result<
+        ExPolicy, hpx::util::tagged_pair<tag::in(InIter), tag::out(OutIter)>
+    >::type
+    transfer(ExPolicy && policy, InIter first, InIter last, OutIter dest)
+    {
+        typedef typename std::iterator_traits<InIter>::iterator_category
+            input_iterator_category;
+        typedef typename std::iterator_traits<OutIter>::iterator_category
+            output_iterator_category;
+
+        static_assert(
+            (boost::is_base_of<
+                std::input_iterator_tag, input_iterator_category>::value),
+            "Required at least input iterator.");
+
+        static_assert(
+            (boost::mpl::or_<
+                boost::is_base_of<
+                    std::forward_iterator_tag, output_iterator_category>,
+                boost::is_same<
+                    std::output_iterator_tag, output_iterator_category>
+            >::value),
+            "Requires at least output iterator.");
+
+        typedef hpx::traits::segmented_iterator_traits<InIter> iterator_traits;
+        typedef typename iterator_traits::is_segmented_iterator is_segmented;
+
+        return hpx::util::make_tagged_pair<tag::in, tag::out>(
+            detail::transfer_<Algo>(
+                std::forward<ExPolicy>(policy), first, last, dest,
+                is_segmented())
+            );
+    }
+}}}
+
```

Comment:
I'd suggest to move the transfer() algorithm into the detail namespace as it is not  to be exposed to users.